tools/cldr/cldr-to-icu/src/main/java/org/unicode/icu/tool/cldrtoicu/localedistance/Trie.java

```diff
@@ -0,0 +1,112 @@
+// © 2017 and later: Unicode, Inc. and others.
+// License & terms of use: http://www.unicode.org/copyright.html#License
+package org.unicode.icu.tool.cldrtoicu.localedistance;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+
+import java.nio.ByteBuffer;
+import java.util.function.Consumer;
+
+import com.ibm.icu.impl.locale.LocaleDistance;
+import com.ibm.icu.util.BytesTrieBuilder;
+
+/**
+ * Trie constructed by adding "spans" of data representing prefix
+ * sequences with mapped values.
+ *
+ * <p>When a prefix needs to be added to a Trie, a new span is created to
+ * represents the additional data. If a final value is added to a span, then
+ * the current prefix data is committed to the underlying Trie as its key.
+ *
+ * <p>Typical use might look like:
+ * <pre>{@code
+ * Trie trie = new Trie();
+ * mappedData.forEach(
+ *     (prefix, subValues) -> trie.root().with(prefix, subSpan -> process(subSpan, subValues));
+ * byte[] bytes = trie.toByteArray();
+ * }
+ * }</pre>
+ * where the {@code process} method may create more sub-spans, and eventually
+ * calls {@link Span#putPrefixAndValue(int)} to commit the current sequence
+ * of prefixes and the given value to the Trie.
+ *
+ * <p>Since spans share a common buffer for prefix data, it is important
+ * that extended spans are consumed before the parent span is used again.
+ * This is one reason why the API requires a consumer to be passed when a
+ * span is extended.
+ */
+final class Trie {
+    private final BytesTrieBuilder trieBuilder = new BytesTrieBuilder();
+    private final byte[] spanBytes = new byte[24];
+
+    /**
+     * Represents a sequence of prefixes to be added to the underlying Trie
+     * when a value is specified.
+     *
+     * <p>The position of a span cannot be modified, but they are not thread
+     * safe (since they share the same underlying buffer).
+     */
+    final class Span {
+        // The index *after* the last prefix was added.
+        private final int index;
+
+        // The root span.
+        private Span() {
+            this.index = 0;
+        }
+
+        // An extended span with the given prefix included.
+        private Span(int index, String prefix) {
+            checkArgument(index >= 0, "bad index: %s", index);
+            checkState(!prefix.isEmpty(), "invalid subtag: %s", prefix);
+            checkState(index + prefix.length() <= spanBytes.length, "span too long");
+            if (prefix.equals("*")) {
+                spanBytes[index++] = '*';
+            } else {
+                checkArgument(!prefix.contains("*"), "prefix must not contain '*': %s", prefix);
+                for (int i = 0; i < prefix.length(); i++) {
+                    char c = prefix.charAt(i);
+                    checkArgument(c < 0x7e, "invalid trie character: %s", c);
```

Comment:
I didn't know about END_OF_SUBTAG.
I picked 7e, because 7f is unprintable and it felt odd to allow it in data which might need debugging via printout. I probably meant <= 7e / < 7f and typoed.
Happy to use END_OF_SUBTAG since it's go a name and docs.